### PR TITLE
Refactor buildpack

### PR DESCRIPTION
**Pull request by nickcatal**
Hey @gregburek & co

I refactored this buildpack to allow better use on docker and get it closer to how the nginx buildpack is done. Here's a summary of the changes:

* Added `bin/start-pgbouncer`, allowing users to bypass stunnel (useful if you want to use this with postgres in docker)
* Included pgbouncer and stunnel binaries in buildpack (less reliance on a 3rd party's s3 bucket)
* Moved to pgbouncer 1.7 and stunnel 5.28 (the latest as of Dec 23, 2015)
* Removed support for legacy cedar platform. Only cedar-14 is supported
* Moved building of pgbouncer and stunnel to heroku itself (using the same setup that the nginx buildpack has to build their binaries)
* Put stunnel and pgbouncer binaries in `bin/` folder, and configs in `config/`
* Changed regex for URLs to support password-less postgres users
* Removed noisy per-client-connection and per-client-disconnection logging default, reducing log size for apps that connect to pgbouncer once per request
* Bumped the changelog to v0.4.0
* Updated the readme and contributor instructions (although if I didn't do that enough let me know)
* Removed the old Dockerfile

Let me know if you see any problems.

Thanks!

Closes #65

### CODE REVIEW

**Comment by gregburek** [comment]
Hello and thank you for the pr. 

I'll respond to each bullet point in line:

> - Added bin/start-pgbouncer, allowing users to bypass stunnel (useful if you want to use this with postgres in docker)

This is not desirable and presents a foot gun to users, as it would be tempting to not use stunnel in a production environment where tls connection protection is required by Heroku Postgres and it represents a best practice. Dev/prod parity should be maintained by using stunnel in all deployments.

> - Removed support for legacy cedar platform. Only cedar-14 is supported

This is fair, as cedar-10 was EOL on November 5, 2015 (https://devcenter.heroku.com/changelog-items/755).

> - Included pgbouncer and stunnel binaries in buildpack (less reliance on a 3rd party's s3 bucket)
> - Moved to pgbouncer 1.7 and stunnel 5.28 (the latest as of Dec 23, 2015)
> - Moved building of pgbouncer and stunnel to heroku itself (using the same setup that the nginx buildpack has to build their binaries)
> - Removed the old Dockerfile

This is a problematic change, as the docker build process is needed in order to develop the patch https://github.com/heroku/heroku-buildpack-pgbouncer/blob/master/support/pgbouncer-1.5.4-sigterm.diff which prevents pgbouncer exiting early as outlined in #18. As the new pgbouncer binary does not have this patch, and no testing is documented as to the new version's exit behavior, the new binaries and new build process are not usable as presented in this pr. 

The use of git to store and transport the stunnel and pgbouncer binaries is a bit icky, but would probably improve reliability as it is possible to get s3 settings wrong, which has happened before.

I will run through the build process with the new binaries to check pgbouncer's exit behavior and see if a patch is needed. Then, I will create a new pr with updated binaries and patches with the existing docker build methods.

> - Put stunnel and pgbouncer binaries in bin/ folder, and configs in config/

This goes against what I understand as a best practice for buildpacks, where inserted binaries and configs are limited to `vendors` dir in an effort to limit hypothetical conflicts and new dirs in a given app.

> - Changed regex for URLs to support password-less postgres users

As mentioned in #64, I question why `trust` is not used or a password set in docker's env. However, the new regex, seems nicer and I'll give it a spin in my testing apps to see if it is able to function properly.

> - Removed noisy per-client-connection and per-client-disconnection logging default, reducing log size for apps that connect to pgbouncer once per request

As mentioned in #63, changing the default to accommodate an edge case would pose a barrier for new users to confirm correct usage of this buildpack. I'll be adding a note to the `README` instead.

> - Bumped the changelog to v0.4.0
> - Updated the readme and contributor instructions (although if I didn't do that enough let me know)

In summary, in a new pr I will:
- test pgbouncer's exit behavior and see if a SIGTERM handling patch is needed
- build new binaries (with patch if needed) and add them to git
- deprecate the cedar-10 behavior
- cherry pick and test new regex that allows for password-less connection strings
- update readme with regarding ephemeral db connections and logging options
- bump changelog to v0.4.0 and cut a new release

Thank you again for the pr and your work on these issues. I would welcome your review once I post the new pr.

**Comment by gregburek** [comment]
I now see that pgbouncer has added TLS support in 1.7 (https://github.com/pgbouncer/pgbouncer/blob/master/NEWS.rst). I'll take a look at dropping stunnel in favor of this.

**Comment by nickcatal** [comment]
For what it's worth, 90% of this was to bring it in line with how the nginx buildpack does it.

>I now see that pgbouncer has added TLS support in 1.7 (https://github.com/pgbouncer/pgbouncer/blob/master/NEWS.rst). I'll take a look at dropping stunnel in favor of this.

That would be a great solution. But you'll still need to support stunnel for legacy users.

Either stunnel has to be dropped from being a requirement or the github docker CLI plugin needs to change what `postgres` image is generated with `docker-compose.yml` to one that supports stunnel.

>This goes against what I understand as a best practice for buildpacks, where inserted binaries and configs are limited to vendors dir in an effort to limit hypothetical conflicts and new dirs in a given app.

I cloned what the nginx buildpack does, since I use that buildpack on every web dyno I also use pgbouncer on. Did you read somewhere that's the best way to do it? If so, I'll change how I do buildpacks. I just couldn't find anything.

>As mentioned in #64, I question why trust is not used or a password set in docker's env. However, the new regex, seems nicer and I'll give it a spin in my testing apps to see if it is able to function properly.

Good point. I added that feature then realized that it was smarter to configure the docker postgres with a password. Feel free to ignore.

>As mentioned in #63, changing the default to accommodate an edge case would pose a barrier for new users to confirm correct usage of this buildpack. I'll be adding a note to the README instead.

Will reply there. But a README update would be useful.

**Comment by gregburek** [comment]
> That would be a great solution. But you'll still need to support stunnel for legacy users.

Is there a particular thing that stunnel provides that the new TLS pgbouncer does not? I notice that the RDS `option = no_ticket` thing is not possible, so I'm going to give it a spin and see if there is a work around or it can be implemented. Once that is figured out, I *think* this is a drop in replacement. Am I missing something?

> Either stunnel has to be dropped from being a requirement or the github docker CLI plugin needs to change what postgres image is generated with docker-compose.yml to one that supports stunnel.

I think I can use `prefer` for server_tls_sslmode to allow for both use of a docker container without TLS and Heroku Postgres which requires it. 

> I cloned what the nginx buildpack does, since I use that buildpack on every web dyno I also use pgbouncer on. Did you read somewhere that's the best way to do it? If so, I'll change how I do buildpacks. I just couldn't find anything.

To be clear, this is my best practice, as I understand it from observing our ruby and go buildpacks, which are more actively maintained than ryan's buildpack.

> Good point. I added that feature then realized that it was smarter to configure the docker postgres with a password. Feel free to ignore.

I took the opportunity to use the `read` parsing and also improve handling of extraneous `?bar=foo` options that are tacked on the end of the string. The previous regex would put the db name and options into $6 and leave $5 blank, which would then rely on whitespace separators to fall through. This is now fixed and connections options are preserved with the new parsing regime.

Please take a look at #67 which implements most of what we have discussed. I'll use another pr to see if stunnel can be replaced with pgbouncer tls.

**Comment by nickcatal** [comment]
>Is there a particular thing that stunnel provides that the new TLS pgbouncer does not? I notice that the RDS option = no_ticket thing is not possible, so I'm going to give it a spin and see if there is a work around or it can be implemented. Once that is figured out, I think this is a drop in replacement. Am I missing something?

The prefix for your `Procfile` is `bin/start-pgbouncer-stunnel`, so if you remove stunnel you'll still need `bin/start-pgbouncer-stunnel` to do something otherwise people's processes will stop working.

And presumably people expect stunnel to still be there?

>I think I can use prefer for server_tls_sslmode to allow for both use of a docker container without TLS and Heroku Postgres which requires it.

You can. I just tested this locally (inside an ubuntu docker container) with both a heroku database and a RDS one and both will connect via TLSv1.2 if you have `server_tls_sslmode` set to `prefer`

So that's indeed a pretty straightforward solution to this problem.

**Comment by gregburek** [comment]
In #68, I've created a symbolic link from `bin/start-pgbouncer-stunnel` to `bin/start-pgbouncer`, which may need to exist in perpetuity to handle legacy deployments. I'm ok with that. 

You can give the new branch a spin by appending `#no-stunnel` to the buildpack invocation.

I was also able to confirm working with Heroku Postgres, but I need to try and reproduce an issue with a sleeping dyno waking up and trying to connect to RDS before this is a slam dunk, I think.
